src/py21cmfast/drivers/param_config.py

```diff
@@ -55,18 +78,48 @@ class InputParameters:
     different sets of instances.
     """
 
-    random_seed = attrs.field(default=None, converter=attrs.converters.optional(int))
-    redshift = attrs.field(default=None, converter=attrs.converters.optional(float))
+    random_seed = attrs.field(converter=int)
     user_params: UserParams = input_param_field(UserParams)
     cosmo_params: CosmoParams = input_param_field(CosmoParams)
     flag_options: FlagOptions = input_param_field(FlagOptions)
     astro_params: AstroParams = input_param_field(AstroParams)
 
+    # These private fields can be used for controlling node_redshifts if an array is not passed explicitly,
+    #   but are not used for comparisons
+    _min_redshift = attrs.field(converter=float, default=6.0, eq=False, repr=False)
+    _max_redshift = attrs.field(
+        converter=float, default=global_params.Z_HEAT_MAX, eq=False, repr=False
+    )
+    _redshift_step = attrs.field(
+        converter=float, default=global_params.ZPRIME_STEP_FACTOR, eq=False, repr=False
+    )
+
+    # passed to the converter, TODO: this can be cleaned up
+    node_redshifts = attrs.field(
+        converter=attrs.Converter(_node_redshifts_converter, takes_self=True)
+    )
```

Comment:
I think this is OK. It's a little convoluted, but does everything we need it to. I think this would be cleaner again if we were to remove Z_HEAT_MAX and ZPRIME_STEP_FACTOR as global variables. In that case, we'd just let the user construct the array every time (and provide a convenience method that they could call to do that easily).